lib/eigensolve_quda.cpp

```diff
@@ -214,6 +216,95 @@ namespace quda
     return sum;
   }
 
+  void EigenSolver::permuteVecs(std::vector<ColorSpinorField *> &kSpace, int *mat, int size)
+  {
+    std::vector<int> pivots(size);
+    for (int i = 0; i < size; i++) {
+      for (int j = 0; j < size; j++) {
+        if (mat[j * size + i] == 1) { pivots[j] = i; }
+      }
+    }
+
+    // Identify cycles in the permutation array.
+    // We shall use the sign bit as a marker. If the
+    // sign is negative, the vector has already been
+    // swapped into the correct place. A positive
+    // value indicates the start of a new cycle.
+
+    for (int i = 0; i < size; i++) {
+      // First cycle always starts at 0, hence OR statement
+      if (pivots[i] > 0 || i == 0) {
+        int k = i;
+        // Identify vector to be placed at i
+        int j = pivots[i];
+        pivots[i] = -pivots[i];
+        while (j > i) {
+          std::swap(kSpace[k + num_locked], kSpace[j + num_locked]);
+          pivots[j] = -pivots[j];
+          k = j;
+          j = -pivots[j];
+        }
+      }
+    }
+    // Sanity check
+    for (int i = 0; i < size; i++) {
+      if (pivots[i] > 0) errorQuda("Error at pivot %d", i);
+    }
+  }
+
+  void EigenSolver::blockRotate(std::vector<ColorSpinorField *> &kSpace, double *array, int rank, int is, int ie,
```

Comment:
For readability/clarity, let's change `int type` into a file-local enum indicating full, lower, upper.

Reply:
39e1a1ab7b8a533573944594f54db6c537305236